generator/src/Google/Service/Generator/Method.php

```diff
@@ -0,0 +1,162 @@
+                }
+            }
+            ksort($method['parameters']);
+            $this->parameters = array_merge($parameters, $method['parameters']);
+
+            foreach ($this->parameters as $k => $v) {
+                if (isset($v['required'])) {
+                    $var_name = '$'. lcfirst(StringUtilities::ucstrip($k));
+                    $this->paramsP[] = $var_name;
+                    $this->paramsA[] = "'$k' => $var_name";
+                    $this->phpdocParams[] = Method::getPhpDocParam($var_name, $v);
+                } else {
+                    $var_name = lcfirst(StringUtilities::ucstrip($k));
+                    $this->phpdocOptParams[] = Method::getPhpDocParam($var_name, $v);
+                }
+            }
+        }
+
+        if (isset($method['request']) && $method['httpMethod'] != 'GET') {
+            $this->requestRef = $method['request']['$ref'];
+            $this->hasPostBody = true;
+        }
+
+        if (isset($method['response'])) {
+            $this->responseRef = $method['response']['$ref'];
+        }
+    }
+
+    public function getParamsParams($modelclassname)
+    {
+        $params = $this->paramsP;
+        if ($this->hasPostBody) {
+            $params[] = "$modelclassname \$postBody";
+        }
+        $params[] = '$optParams = array()';
+        return implode(", ", $params);
+    }
+
+    public function getParamsArray()
+    {
+        $params = $this->paramsA;
+        if ($this->hasPostBody) {
+            $params[] = "'postBody' => \$postBody";
+        }
+        return implode(", ", $params);
+    }
+
+    public function getDescription()
+    {
+        $tmparr = explode('.', $this->id);
+        array_shift($tmparr);
+        $desc = '('. implode('.', $tmparr) . ')';
+        if ($this->description != '') {
+            $desc = "$this->description $desc";
+        }
+        return StringUtilities::commentWordwrap($desc);
+    }
+
+    public static function getPhpDocParam($name, $node)
+    {
+        if (isset($node['required'])) {
+            $token = '@param';
+
+            if (isset($node['repeated'])) { // TODO Why is this only for required parameters???
+                $repeatToken = '|array';
+            } else {
+                $repeatToken = '';
+            }
+        } else {
+            $token = '@opt_param';
+            $repeatToken = '';
+        }
+
+        if (isset($node['type'])) {
+            $typeToken = Method::PHP_TYPES[$node['type']];
+        }
+        if (isset($node['format'])
+            && ($node['format'] == 'uint32' || $node['format'] == 'uint64')
+        ) {
+            $typeToken = "string";
+        }
+        if (isset($node['description'])) {
+            $desc = $node['description'];
+        } else {
+            $desc = '';
+        }
+        $str = "$token $typeToken$repeatToken $name $desc";
+        return StringUtilities::commentWordwrap($str);
+    }
+
+    const PHP_TYPES =
```

Comment:
Would we be able to move constant definitions to the top of the file please?

Reply:
☑️